### PR TITLE
add isSupported field to modules

### PR DESCRIPTION
**Pull request by jakemac53**
Towards https://github.com/dart-lang/build/issues/1810

Use the `sdkDeps` and `platform.supportsLibrary` fields to check for module support, and encode that in the module.

### CODE REVIEW

File: build_modules/lib/src/meta_module.dart [hunk 1]
```diff
@@ -52,7 +49,9 @@ Module _moduleForComponent(
   var directDependencies = Set<AssetId>()
     ..addAll(componentLibraries.expand((n) => n.depsForPlatform(platform)))
     ..removeAll(sources);
-  return Module(primaryId, sources, directDependencies, platform);
+  var isSupported = componentLibraries
+      .every((l) => l.sdkDeps.every((d) => platform.supportsLibrary(d)));
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_modules/lib/src/meta_module.dart]: Actually surprised the lint didn't catch this, updated

File: build_modules/test/meta_module_builder_test.dart [hunk 2]
```diff
@@ -26,4 +27,62 @@ main() {
...
+        'a|lib/a.module.library':
+            ModuleLibrary.fromSource(assetA, "import 'src/b.dart';")
+                .serialize(),
+        'a|lib/src/b.module.library':
+            ModuleLibrary.fromSource(assetB, "import 'dart:io';").serialize(),
+      }, outputs: {
+        'a|lib/${metaModuleExtension(DartPlatform.dart2js)}':
+            encodedMatchesMetaModule(metaA),
+      });
+    });
```
Comment (samogot) [inline build_modules/test/meta_module_builder_test.dart]: Maybe add a test to show that `isSupported` doesn't bubbles up transitively between modules?

File: build_modules/lib/src/modules.dart [hunk 3]
```diff
@@ -96,6 +96,16 @@ class Module {
...
 
+  /// Whether or not this module is supported for [platform].
+  ///
+  /// Compilers can use this to either silently skip compilation of this module
+  /// or throw early errors or warnings.
+  ///
+  /// Modules are allowed to exist even if they aren't supported, which can help
+  /// with discovering root causes of incompatibility.
+  @JsonKey(name: 'is', nullable: false)
+  final bool isSupported;
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_modules/lib/src/modules.dart]: done

File: build_modules/test/meta_module_builder_test.dart (hunk 2 above)
Comment (jakemac53) [reply build_modules/test/meta_module_builder_test.dart]: done

**Pull request merged**